### ISSUE TITLE
Provide access to arguments of parameterized tests in lifecycle callback methods

### ISSUE THREAD

**Issue by larryricker** [junit-team/junit5#944]
## Overview

Simple parameter resolution for strings, ints, doubles, longs works in M4, but no longer works in M5 and throws the following exception. 

```
ParameterResolutionException: No ParameterResolver registered for parameter [java.lang.String arg0] in executable
```

Was previously using M4, upgraded to M5 and issue began occurring.

Example:

```java
@ParameterizedTest (name = "{index} [{arguments}] User Flags CONDITION NAME")
@ValueSource(strings = { "dev", "sit" })
@DisplayName("{index} [{arguments}] User Flags CONDITION NAME")
public void testUseTestScenario(String testEnvironment, TestInfo info, TestReporter testReporter) throws Exception {
}
```

Do not get a build exception, get a runtime exception with jUnit5 with gradle 3.5 running in Jenkins.

```
    => org.junit.jupiter.api.extension.ParameterResolutionException: No ParameterResolver registered for parameter [java.lang.String arg0] in executable [public void com.orgname.TestSituation.setUpBeforeEach(java.lang.String,org.junit.jupiter.api.TestInfo,org.junit.jupiter.api.TestReporter) throws java.lang.Exception].
  JUnit Jupiter:TestSituation:User Flags Situation A:2 [sit] User Flags Situation A
    MethodSource [className = 'com.orgname.TestSituation', methodName = 'testSituationA', methodParameterTypes = 'java.lang.String, org.junit.jupiter.api.TestInfo, org.junit.jupiter.api.TestReporter']
    => org.junit.jupiter.api.extension.ParameterResolutionException: No ParameterResolver registered for parameter [java.lang.String arg0] in executable [public void com.orgname.TestSituation.setUpBeforeEach(java.lang.String,org.junit.jupiter.api.TestInfo,org.junit.jupiter.api.TestReporter) throws java.lang.Exception].
```

## Related Issues

- #833 
- #1139
- #1668
- #1884
- #4352

## Deliverables

- [ ] ...

**Comment by sbrannen** [comment]
FYI: this is a direct result of the changes performed in conjunction with #833.

**Comment by sbrannen** [comment]
Assigned to 5.0 M6 in order to make a decision about how to proceed.

**Comment by marcphilipp** [comment]
The reason is that we no longer resolve parameters from `@Source` annotations for `@BeforeEach` etc. methods. Do you need the parameter there?

**Comment by larryricker** [comment]
Yes.  I need the parameter on @BeforeEach and @BeforeAfter for several tests for Selenium and Appium testing.

**Comment by sormuras** [comment]
@larryricker Look at how Paul is solving the WebDriver injection task: https://github.com/paul-hammant/JUnit5_DependencyInjection_WebDriver

**Comment by Rendaw** [comment]
If I've understood the topic correctly, I need this as well.

I have a dozen or so complex objects that for the environment used for numerous tests.  I recently identified a couple default parameters for the environment setup that need to be tested at multiple values.  My only options seem to be explicitly forwarding parameters to a setup function in each test case or downgrading to M4.

Also, it would be nice if it somehow worked with BeforeAll so heavy immutable fixtures don't get generated and destroyed at each test method.

**Comment by opncow** [comment]
Any news on this? Just migrated to JUnit 5 for the improved parameterized tests. "Forwarding parameters to a setup function in each test case" feels more like a step back, though. :-/

**Comment by marcphilipp** [comment]
IMHO a setup function that gets parameterized test arguments only makes sense when all test methods in a class are parameterized and use at least similar parameters. Thus, I think it would make sense to support this when the class is parameterized instead of the method (as we plan to do in #878).

**Comment by sbrannen** [comment]
I agree with @marcphilipp's analysis.

**Comment by Tradunsky** [comment]
Reproduces on 5.1.1 and on the latest 5.2.0-M1

org.junit.jupiter.api.extension.ParameterResolutionException: No ParameterResolver registered for parameter [java.lang.String arg1] in executable 

@ParameterizedTest(name = "meaningful name and argument {3} based on argument {0} and argument {1}")
    @CsvSource({
            "6", "10", "40",
            "1", "10", "90"
    })
    public void meaningfulTestName(String meaningfulArg1, String meaningfulArg2, String meaningfulArg3){

The same with long, int

**Comment by sbrannen** [comment]
@Tradunsky, your stack trace is incomplete: `in executable`. 

Which executable?

If it's for a `@BeforeEach`/`@AfterEach` method, then that's expected since this behavior has not changed since 5.0 GA.

**Comment by geo-m** [comment]
@Tradunsky 
Shouldnt it be "6, 10, 40" instead of "6","10","40" ?

**Comment by Tradunsky** [comment]
@geo-m Good catch! Works like a charm.
Thanks.

@sbrannen 
Thank you as well

**Comment by sbrannen** [comment]
Indeed, good catch, @geo-m!

**Comment by scharfstein-dev** [comment]
I have a somewhat convoluted but (I think) valid scenario for needing parameter injection outside of a test method - either in the constructor or a `@BeforeEach` setup method.

My context:
In our application, we use scenario tests to reproduce situations that have led to errors in the past. Such a scenario is associated with certain test data, configuration parameters, and so on. All test classes wishing to use a certain scenario must extend a specific base class which arranges setup for the scenario (like booting up the application, configuring it and feeding it the test data). The base class performs this task in a `@BeforeEach` method.

What I am trying to do:
The scenario configurations mentioned above need to fulfill certain consistency criteria. If they are entered incorrectly, this can introduce hard-to-find test errors. The consistency criteria can be checked automatically, and that's what I'm trying to do here. Basically, I am implementing a custom scenario test for the scenario I want to check, then I am manipulating the configuration during setup (also in a `@BeforeEach` method). Then, if the base class setup throws an error due to detecting the inconsistent data, the consistency test passes. If the manipulation is not detected however, the original configuration must already have been inconsistent, and the user is alerted to this with a failing test.

I want to do this for all existing scenarios automatically, so I've written a parameterized test for it that does exactly this. Works very well for me with JUnit 4, but if I want to migrate it I need to know the scenario configuration in the mentioned `@BeforeEach` method so I can manipulate it. Later on (in the `@Test` method) would be too late, because the base class has already completed setup by this point.

I know this must be a bit confusing, but I hope I have been clear enough in my description. If you have any clarification requests, please let me know!

**Comment by sbrannen** [comment]
Hi @scharfstein-dev,

> I know this must be a bit confusing, but I hope I have been clear enough in my description. If you have any clarification requests, please let me know!

I don't think that's especially confusing; however, I do wonder if parameterized test classes wouldn't be a better solution to your problem.

Have you seen #871 and #878?

**Comment by sormuras** [comment]
> Have you seen #871 and #878?

Target _parameterized test classes_ for **5.4.0-M2**?

**Comment by tobiasdiez** [comment]
This is still not possible, and would be helpful indeed.

**Comment by zcmgyu** [comment]
In my case, I want to connect and create WebDriver to multi-devices at beforeEach, to run test cases in parallel.

**Comment by Jigar2018** [comment]
Hey guys! New to junit and doing migration work from junit 4 to junit 5. I asked [this question](https://github.com/junit-team/junit5/issues/2872) but got to know it's kind of a duplicate to this open item. Anyone can help me understand how I can achieve this? (Please check sample code [here](https://github.com/junit-team/junit5/issues/2872)). Thanks team!

**Comment by KeatsPeeks** [comment]
It's rather infortunate that many years in, Junit 5 still can't do everything that JUnit 4 could. 

You can work around this by calling a setup method manually, or creating a (complex) ParameterResolver.

**Comment by wangyihong-yvonne** [comment]
Any instruction on how to create ParameterResolver?

**Comment by sbernard31** [comment]
I agree Parameterized class feature with parameter in @BeforeEach @AfterEach would be really useful. :+1: 

> Any instruction on how to create ParameterResolver?

You could look at : https://stackoverflow.com/a/69265907/5088764

**Comment by mirza-suljic** [comment]
I came across this issue by looking for a way to make a parametrized `@BeforeAll`. I'd definately appreciate this.

**Comment by uchagani** [comment]
I'll add another use case for this:  I need to get the value of the parameter from a `@ParameterTest` so that I can do some clean up using that value.

**Comment by t1** [comment]
Have you any thoughts about the possible API? I think coming up with a good one is not as easy as one would hope, because a `ParameterResolver` can resolve a parameter by type, index, annotations, parameter name (if the code is compiled with the `-parameters` option), or anything that it could derive from the test instance or `ExecutionContext`, e.g. tags. But the API we need here has to work the other way around; and it's not easy to reverse such a lookup.

Maybe we can add a simple method that allows only backward resolution by type: `<T> T ExtensionContext#getResolvedParameterByType(Class<T>)`. An extension always needs the type of a parameter, so this is the most minimalistic approach I can think of. While very limited, it is probably also the most robust way for extensions to find a parameter: the resolvers can change their means of operation, but as long as there is only one resolution for a type, test extensions can still fetch them. This obviously won't work for, e.g., string parameters, but I'm among those who try to avoid "stringly typed" programming anyway. And it's always possible to wrap such data in, e.g., a single-component record.

I think this would make a fine first step that helps most extension writers. And if it's not sufficient for other use-cases, we can later think about the details of other means of backward resolution.

**Comment by marcphilipp** [comment]
> Have you any thoughts about the possible API?

Wouldn't it be sufficient to get access to an array or list of objects via `ExtensionContext`?

**Comment by t1** [comment]
> Wouldn't it be sufficient to get access to an array or list of objects via ExtensionContext?

You mean something like `List<?> ExtensionContext#getResolvedParameters()` then you can filter yourself? Somewhat basic, but a very good first step.

Maybe I can find the time to create a PR.

**Comment by marcphilipp** [comment]
It might be worth returning a custom interface so we can add convenience lookup methods to it later rather than adding more methods to `ExtensionContext`:

```java
interface ExtensionContext {
	// ...
	Optional<ResolvedArguments> getResolvedArguments();
}
interface ResolvedArguments {
	List<?> getRawArguments();
	// we can add lookup methods here later
}
```

**Comment by marcphilipp** [comment]
@t1 What was your use case for this enhancement? In which lifecycle methods or extension methods would you want to access the resolved parameters? Resolution happens just before the invocation of the test method so no lifecycle methods or callbacks that are invoked _before_ the invocation would be able to access them.

**Comment by marcphilipp** [comment]
**Team epiphany:** There's already a way to access the arguments of test methods. An extension can implement `InvocationInterceptor`, override `interceptTestMethod`, and call the `getArguments` method on the supplied `ReflectiveInvocationContext`.

**Comment by danielhodder** [comment]
The main reason I would want this is to be able to use extensions (like `TemporryFolder`, or an extension that creates a database) in my `@Before` setup methods. A really basic example would be something like this:

```java
@ExtendWith(CreateDatabase.class)
class SchemaTest {
    @Before
    public void setup(javax.sql.Datasource db) {
        // Create data here
    }
    
    @Test
    public void test(javax.sql.Datasource db) {
        // Verify data in the database
    }
}
```

**Comment by t1** [comment]
@marcphilipp I have moved on to the next project and already completely forgot about this issue and the use-case behind it... my memory is terrible 😳. But I went back and created a reproducer, just to find out that... you are completely right: instead of using a `BeforeEachCallback`, I can use a `TestTemplateInvocationContextProvider`. Works like a charm. Many thanks!

**Comment by t1** [comment]
Just in case anybody is interested, my reproducer/demo is here: https://github.com/t1/junit-testtemplate-invocationinterceptor-demo

**Comment by marcphilipp** [comment]
@danielhodder That's already possible by implementing a `ParameterResolver` that provides the same `javax.sql.Datasource` for both methods and uses `ExtensionContext.Store` to cache it and automatically close it (via `CloseableResource`), if necessary. There's an example in the [User Guide](https://junit.org/junit5/docs/current/user-guide/#extensions-keeping-state).

**Comment by marcphilipp** [comment]
> you are completely right: instead of using a `BeforeEachCallback`, I can use a `TestTemplateInvocationContextProvider`. Works like a charm. Many thanks!

You mean `InvocationInterceptor`, right?

**Comment by t1** [comment]
@marcphilipp: yes. sorry for the confusion.

**Comment by marcphilipp** [comment]
To summarize, it's already possible to access the arguments of any test method, including parameterized ones, from an extension by implementing `InvocationInterceptor`, overriding the appropriate method (for example, `interceptTestTemplateMethod`), and calling the `getArguments` method on the supplied `ReflectiveInvocationContext`.

In general, parameter resolution happens immediately before invoking the method. Therefore, arguments cannot be provided to lifecycle callback methods that are invoked _before_ (for example, `@BeforeEach`) the (parameterized) test method is called. It would be possible to write an extension that stores them and provides them to lifecycle methods that are invoked _after_ (for example, `@AfterEach`) the (parameterized) test method.

However, in the specific case of `@ParameterizedTest` methods, the set of arguments for an invocation is already known prior to invoking lifecycle methods. Thus, it could be made available to them. Since different `@ParameterizedTest` method in a test class often have different parameters, re-declaring the parameters directly on lifecycle methods is brittle. We could support injecting `ArgumentAccessor` but, as with re-declaring the parameters, this would fail if there's a `@Test` method in the same test class. It seems to me that this use case would be better served by specifying parameters on the class level (which we'll add in #878) which will inject them into the test class constructor or its fields from where lifecycle methods can access them.

**Comment by t1** [comment]
👍

Minor addition: the same is also true for `@TestTemplate` tests.

**Comment by marcphilipp** [comment]
We could base this work on #4370 and support injecting a `ParameterInfo` object into lifecycle and other methods. Writing a `ParameterResolver` that does that based on the PR would be straightforward. However, in the case of nested parameterization, for example, it would be confusing which level gets injected.

```java
@ParameterizedClass
@ValueSource(ints = { 1, 2 })
record SomeTests(int number) {

	@BeforeEach
	void beforeEach(ParameterInfo parameterInfo) {
		// Would contain info for `parameterizedTest` when called for its invocations
        // Would contain info for `SomeTests` when called for regular `@Test` methods
	}

	@ParameterizedTest
	@ValueSource(strings = { "foo", "bar" })
	void parameterizedTest(String text) {/*...*/}

	@Test
	void regularTest() {/*...*/}
}
```

This could be even more nested in case `parameterizedTest` is declared inside a `@Nested` `@ParameterizedClass`.

Any ideas how to solve that?

**Comment by marcphilipp** [comment]
**Team decision:** We think many if not all of the above use cases are better served by using `@ParameterizedClass` (from #878) or writing a custom extension. In light of that, we're closing this issue. However, if you think there is a use case that can't be addressed in that way, please let us know.
